#### Q8. m = 8일 때 p = 0.1, 0.5, 0.8 각각의 이항 분포를 그림으로 그리시오.  


In [9]:
import numpy as np
from math import factorial
import plotly.graph_objects as go
from plotly.subplots import make_subplots


# Probability density of the binomial distribution
def bin_dist(k, n, p):
    nck = factorial(n) / (factorial(k) * factorial(n - k))
    pd = nck * p**k * (1-p)**(n-k)
    return pd

x = np.arange(9)

n = 8 # 시행횟수
prob = [0.1, 0.5, 0.8] #n번 시행했을 때 확률값, 10%, 50%, 80%

row = [1, 1, 2]
col = [1, 2, 1]

fig = go.Figure()

for i in range(len(prob)):
    pd = np.array([bin_dist(k, n, prob[i]) for k in range(n+1)])

fig = make_subplots(
    rows=2, cols=2,
    subplot_titles=("n, p = 8, 0.1", "n, p = 8, 0.5", "n, p = 8, 1.5", "Total"),
    shared_xaxes=True,
    shared_yaxes=True,
    )

for i in range(len(prob)):
    pd = np.array([bin_dist(k, n, prob[i]) for k in range(n+1)])

    fig.add_trace(
        go.Bar(x=x, y=pd),
        row=row[i], col=col[i]
    )

    fig.add_trace(
        go.Scatter(x=x, y=pd, mode='lines'),
        row=2, col=2
    )

# fig.update_xaxes(title_text='k')
# fig.update_yaxes(title_text='P(k)', range=[0, 1])

fig.show() 

#### Q9. 윷놀이에서 P(a) = 0.4, 0.5, 0.6인 3가지 경우에 대해 답하시오.

##### (1) 각 경우의 확률 분포를 구하시오.


In [51]:
from math import factorial
import numpy as np
import plotly.graph_objects as go
from plotly.subplots import make_subplots

def bin_dist(k, n, p):
    ncr = factorial(n) / (factorial(k) * factorial(n - k))
    pd = ncr * p**k * (1-p)**(n-k)
    return pd

case_list = ['A', 'B', 'C', 'D']
title_list = ['윷', '걸', '개', '도', '모']
num = 5

x = np.arange(num)

prob = [0.4, 0.5, 0.6] #n번 시행했을 때 확률값, 10%, 50%, 80%

row = [1, 1, 2, 2]
col = [1, 2, 1, 2]

fig = go.Figure()

fig = make_subplots(
    rows=2, cols=2,
    subplot_titles=("p(a) = 0.4", "p(a) = 0.5", "p(a) = 0.6"),
    shared_xaxes=True,
    shared_yaxes=True,
    )


for p_i in range(len(prob)):
    for j in range(num):
        
        # print(bin_dist(j, len(case_list), prob[p_i]))
        pd = np.array(bin_dist(j, len(case_list), prob[p_i]))
        
        fig.add_trace(
            go.Bar(x=x, y=pd, name=title_list[j]),
            row=row[p_i], col=col[p_i]
        )


fig.show()

##### (2) 각 경우의 엔트로피를 구하시오.

In [6]:
from math import factorial
import numpy as np
import plotly.graph_objects as go
from plotly.subplots import make_subplots

def bin_dist(k, n, p):
    ncr = factorial(n) / (factorial(k) * factorial(n - k))
    pd = ncr * p**k * (1-p)**(n-k)
    return pd

case_list = ['A', 'B', 'C', 'D']
title_list = ['윷', '걸', '개', '도', '모']

prob = [0.4, 0.5, 0.6] #n번 시행했을 때 확률값, 10%, 50%, 80%


for p_i in range(len(prob)):
    hx = 0
    for j in range(len(title_list)):
        pd = np.array(bin_dist(j, len(case_list), prob[p_i]))
        hx -= pd*np.log2(pd)

    print(f"p = {prob[p_i]}일 경우 엔트로피:", hx)
        

p = 0.4일 경우 엔트로피: 1.9920393375694432
p = 0.5일 경우 엔트로피: 2.0306390622295662
p = 0.6일 경우 엔트로피: 1.9920393375694427


[solution]   
자기 정보가 특정 사건 $e_i$ 의 정보량을 측정하는 반면, 엔트로피는 확률분포의 무질서도 또는 불확실성을 측정.

$$H(x) = -\sum P(e_i)\log_{2}(P(e_i))$$

##### (3) 

* p(k)=0.4 와 p(k)=0.5의 교차 엔트로피와 p(k)=0.4와 p(k)=0.6의 교차 엔트로피를 구하시오. 어느 것이 큰지 확인하고 그 이유를 설명하시오. 
* p(k)=0.4, p(k)=0.5의 KL 다이버전스와 p(k)=0.4와 p(k)=0.6의 KL 다이버전스를 구하시오. 어느 것이 큰지 확인하고 그 이유를 설명하시오. 

In [25]:
from math import factorial
import numpy as np
import plotly.graph_objects as go
from plotly.subplots import make_subplots

def bin_dist(k, n, p):
    ncr = factorial(n) / (factorial(k) * factorial(n - k))
    pd = ncr * p**k * (1-p)**(n-k)
    return pd

case_list = ['A', 'B', 'C', 'D']
title_list = ['윷', '걸', '개', '도', '모']

prob = [[0.4, 0.5], [0.4, 0.6]]

for p_i in range(len(prob)):
    hx = 0
    kl = 0

    for j in range(len(title_list)):
        pd = np.array(bin_dist(j, len(case_list), prob[p_i][0]))
        qd = np.array(bin_dist(j, len(case_list), prob[p_i][1]))

        hx -= pd*np.log2(qd)
        kl += pd*np.log2(pd/qd)

    print(f"H{prob[p_i]}일 경우 교차 엔트로피:", hx)
    print(f"kl({prob[p_i][0]} || {prob[p_i][1]})일 경우 교차 엔트로피:", kl)
    print("==========")

H[0.4, 0.5]일 경우 교차 엔트로피: 2.1082369597507684
kl(0.4 || 0.5)일 경우 교차 엔트로피: 0.11619762218132547
H[0.4, 0.6]일 경우 교차 엔트로피: 2.4600093381463672
kl(0.4 || 0.6)일 경우 교차 엔트로피: 0.46797000057692456


[Solution]  
교차 엔트로피와 KL 다이버전스는 비례한다.   
따라서, p(0.4 || 0.6)의 교차 엔트로피, KL 디바이전스가   
p(0.4 || 0.5)의 교차 엔트로피, KL 디바이전스보다 크다. 
KL 디바이전스는 p와 q의 분포의 차이가 멀어질수록 커진다. 

#### 10. 특징 벡터가 3차원이면 퍼셉트론은 매개변수 4개를 가진다. [그림 2-4(a)]가 이 경우에 해당한다. 모든 매개변수가 [-1, 1] 범위에 있고 이 범위를 0.0001 간격으로 나누어 탐색한다고 가정한다.  

(1) 탐색해야 하는 점의 개수를 쓰시오. 점 하나를 평가하는 데 0.01밀리초가 걸린다면, 학습에 걸리는 시간은? 


In [8]:
start = -1
end = 1
sum = -1
step = 0.0001
loop = -1

while sum < end:
  sum += step
  loop += 1

param_num = 4
total = pow(loop, param_num)
learning_time = total*0.01*0.001
learning_time_min = learning_time/60
learning_time_hour = learning_time_min/60
learning_time_day = learning_time_hour/24
learning_time_month = learning_time_day/30
learning_time_year = learning_time_month/30

print("learning_time: ", learning_time)
print("learning_time_min: ", learning_time_min)
print("learning_time_hour: ", learning_time_hour)
print("learning_time_day: ", learning_time_day)
print("learning_time_month: ", learning_time_month)
print("learning_time_year: ", learning_time_year)

learning_time:  1600000000000.0
learning_time_min:  26666666666.666668
learning_time_hour:  444444444.4444445
learning_time_day:  18518518.51851852
learning_time_month:  617283.9506172839
learning_time_year:  20576.131687242796



[solution]
Exhaustive Algorithm을 사용하였을 때, 한 매개변수 당 탐색해야 하는 점의 개수는 20000개이고, 총 4개의 매개변수를 가지므로, 0.0001 간격으로 나누게 된다면 하나의 매개변수가 가질 수 있는 경우의 수는 2 x 10^4 이다.

따라서 네 개의 매개 변수가 가질 수 있는 경우의 수는 (2 x 10^4)^4 = 2 x 10^16 이며, 이를 학습하는데 걸리는 시간은 2 x 10^16 초, 대략 20576년이 걸린다.

(2) [그림 1-6]에 있는 Wine 데이터베이스의 특징 벡터는 13차원이다. 앞의 가정을 그대로 적용한다면 Wine 데이터베이스를 학습할 때 점의 개수와 학습에 걸리는 시간을 쓰시오. 

In [9]:
param_num = 14
total = pow(loop, param_num)
learning_time = total*0.01*0.001
learning_time_min = learning_time/60
learning_time_hour = learning_time_min/60
learning_time_day = learning_time_hour/24
learning_time_month = learning_time_day/30
learning_time_year = learning_time_month/30

print("learning_time: ", learning_time)
print("learning_time_min: ", learning_time_min)
print("learning_time_hour: ", learning_time_hour)
print("learning_time_day: ", learning_time_day)
print("learning_time_month: ", learning_time_month)
print("learning_time_year: ", learning_time_year)

learning_time:  1.6384000000000003e+55
learning_time_min:  2.730666666666667e+53
learning_time_hour:  4.5511111111111114e+51
learning_time_day:  1.8962962962962963e+50
learning_time_month:  6.320987654320988e+48
learning_time_year:  2.1069958847736627e+47


[solution]
매개변수가 14개 이다. 
탐색해야 하는 점의 개수는 1.6384000000000003e+55이다. 
한 매개변수 당 탐색해야 하는 점의 개수는 20000개이고, 총 14개의 매개변수를 가진다.
이를 학습하는데 걸리는 시간은 1.6384000000000003e+55초가 걸린다.